miri engine: Hooks for basic stacked borrows

RalfJung opened the pull request:
r? @oli-obk

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 5b4c5de5f3c4a537d5dd8fc074e5619a47fd158c has been approved by `oli-obk`

<!-- @bors r=oli-obk 5b4c5de5f3c4a537d5dd8fc074e5619a47fd158c -->

RalfJung commented:
@bors r=oli-obk

(I had pushed something here but force-pushed it away again.)

bors commented:
:pushpin: Commit 5b4c5de5f3c4a537d5dd8fc074e5619a47fd158c has been approved by `oli-obk`

<!-- @bors r=oli-obk 5b4c5de5f3c4a537d5dd8fc074e5619a47fd158c -->

RalfJung commented:
@oli-obk I pushed some more commits, matching the [`stacked-borrows`](https://github.com/RalfJung/miri/commits/stacked-borrows) miri branch where the test suite all passes. :)

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/442680832) ([raw log](https://api.travis-ci.org/v3/job/442680832/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

[00:03:53] travis_fold:start:tidy
travis_time:start:tidy
tidy check
[00:03:54] tidy error: /checkout/src/librustc_mir/interpret/eval_context.rs:503: line longer than 100 chars
[00:03:54] tidy error: /checkout/src/librustc_mir/interpret/eval_context.rs:515: line longer than 100 chars
[00:03:55] some tidy checks failed
[00:03:55] 
[00:03:55] 
[00:03:55] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:03:55] 
[00:03:55] 
[00:03:55] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:03:55] Build completed unsuccessfully in 0:00:45
[00:03:55] Build completed unsuccessfully in 0:00:45
[00:03:55] Makefile:79: recipe for target 'tidy' failed
[00:03:55] make: *** [tidy] Error 1

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:2d667194
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
travis_time:end:04603500:start=1539782607036751210,finish=1539782607042283803,duration=5532593
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:00d0ac74
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:08029a38
travis_time:start:08029a38
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:0ea28513
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

oli-obk commented on src/librustc_mir/interpret/cast.rs:
```diff
@@ -47,7 +46,19 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
                 let src = self.read_value(src)?;
-                if self.type_is_fat_ptr(src_layout.ty) {
+
+                if src.layout.ty.is_region_ptr() && dest.layout.ty.is_unsafe_ptr() {
```
slightly unhappy about code that isn't needed during const eval but still will be compiled into const eval. Not sure what to do about that though

RalfJung replied:
I cannot make it depend on `enforce_validity()` as that whitelists some functions in miri and I think my model might go crazy if references are only sometimes tracked.

I could add another `const REF_TRACKING_HOOKS: bool` though.

oli-obk replied:
Good idea, not a blocker for this PR though

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit eb5c69b33c30bca5ebb954caae337c33748368eb has been approved by `oli-obk`

<!-- @bors r=oli-obk eb5c69b33c30bca5ebb954caae337c33748368eb -->

oli-obk commented:
@bors r+

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #55134

bors commented:
:pushpin: Commit eb5c69b33c30bca5ebb954caae337c33748368eb has been approved by `oli-obk`

<!-- @bors r=oli-obk eb5c69b33c30bca5ebb954caae337c33748368eb -->

RalfJung commented:
Fixed your nit, and included https://github.com/solson/miri/pull/487 so that the miri submodule keeps working.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 4e784583aab5ec6ff083e47cb0c3d7fa3518d330 has been approved by `oli-obk`

<!-- @bors r=oli-obk 4e784583aab5ec6ff083e47cb0c3d7fa3518d330 -->

bors commented:
:umbrella: The latest upstream changes (presumably #55171) made this pull request unmergeable. Please resolve the merge conflicts.

RalfJung commented:
I fixed the conflict, and also added a [tiny change](https://github.com/rust-lang/rust/pull/55125/commits/964a20ddc205a4ede847c803e9471e7ffdf83c53) that bugged me after sleeping over this, and the corresponding miri update.

oli-obk commented on src/librustc_mir/interpret/cast.rs:
```diff
@@ -37,37 +37,34 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
-                    src.layout.ty.is_region_ptr() && dest.layout.ty.is_unsafe_ptr()
-                {
+                let src = if M::ENABLE_PTR_TRACKING_HOOKS && src_layout.ty.is_region_ptr() {
+                    assert!(dest.layout.ty.is_unsafe_ptr());
```
add a comment mentioning that the only `Misc` casts that start out with a reference must be to a raw pointer. Clear in hindsight, but took me a minute to figure out why this assert is dead code barring bugs (which is exactly what it's trying to catch)

oli-obk commented:
r=me with comment

RalfJung commented:
@bors r=oli-obk

bors commented:
:pushpin: Commit 50d5cde2dd6b1ae80c170ff873453699d061278b has been approved by `oli-obk`

<!-- @bors r=oli-obk 50d5cde2dd6b1ae80c170ff873453699d061278b -->

bors commented:
:hourglass: Testing commit 50d5cde2dd6b1ae80c170ff873453699d061278b with merge c90f034f45dff5d7d4865259634139ca1cc03bf1...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/19662959)

RalfJung commented:
Fixed by disabling a new miri test on Windows and macOS, where I cannot test them.

This can be much easier debugged on the miri side once things landed in rustc.

@bors r=oli-obk

bors commented:
:pushpin: Commit 3f5b55029632b840169717cd8c627191fe573553 has been approved by `oli-obk`

<!-- @bors r=oli-obk 3f5b55029632b840169717cd8c627191fe573553 -->

bors commented:
:hourglass: Testing commit 3f5b55029632b840169717cd8c627191fe573553 with merge 3779b93f1bdcfa9c7ded680a7b2d715a24c688da...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/444268411?utm_source=github_status&utm_medium=notification)

rust-highfive commented:
The job `x86_64-gnu-aux` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/444268449) ([raw log](https://api.travis-ci.org/v3/job/444268449/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[01:42:38] [TIMING] ToolBuild { compiler: Compiler { stage: 0, host: "x86_64-unknown-linux-gnu" }, target: "x86_64-unknown-linux-gnu", tool: "cargotest", path: "src/tools/cargotest", mode: ToolBootstrap, is_optional_tool: false, source_type: InTree, extra_features: [] } -- 1.158
[01:42:39] testing https://github.com/servo/webrender
[01:42:39] Initialized empty Git repository in /checkout/obj/build/ct/webrender/.git/
[01:42:39] fatal: Could not parse object '57250b2b8fa63934f80e5376a29f7dcb3f759ad6'.
[01:42:59] fatal: unable to access 'https://github.com/servo/webrender/': Could not resolve host: github.com
[01:42:59] thread 'main' panicked at 'assertion failed: status.success()', tools/cargotest/main.rs:128:13
[01:42:59] 
[01:42:59] 
[01:42:59] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/cargotest" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build/ct"
[01:42:59] expected success, got: exit code: 101
[01:42:59] expected success, got: exit code: 101
[01:42:59] 
[01:42:59] 
[01:42:59] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/test/pretty src/test/run-pass/pretty src/test/run-fail/pretty src/test/run-pass-valgrind/pretty src/test/run-pass-fulldeps/pretty src/test/run-fail-fulldeps/pretty src/tools/cargo src/tools/cargotest
[01:42:59] Build completed unsuccessfully in 0:42:57
[01:42:59] make: *** [check-aux] Error 1
[01:42:59] Makefile:60: recipe for target 'check-aux' failed

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:002c04c2
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
travis_time:end:08ffb750:start=1540104389904852277,finish=1540104389914872150,duration=10019873
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0096690f
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:1be205b0
travis_time:start:1be205b0
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:18508c29
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

kennytm commented:
@bors retry travis-ci/travis-ci#9696

bors commented:
:hourglass: Testing commit 3f5b55029632b840169717cd8c627191fe573553 with merge 66910ba686e1e89ff6cfbb46b3e5394bedbe5564...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/19670516), [status-travis](https://travis-ci.org/rust-lang/rust/builds/444309314?utm_source=github_status&utm_medium=notification)
Approved by: oli-obk
Pushing 66910ba686e1e89ff6cfbb46b3e5394bedbe5564 to master...
